File: app/src/main/java/org/bspb/smartbirds/pro/content/MonitoringManager.java
package org.bspb.smartbirds.pro.content;

import static android.content.ContentUris.parseId;
import static org.androidannotations.annotations.EBean.Scope.Singleton;
import static org.bspb.smartbirds.pro.content.Monitoring.Status.paused;
import static org.bspb.smartbirds.pro.content.Monitoring.Status.wip;
import static org.bspb.smartbirds.pro.tools.Reporting.logException;
import static java.lang.Double.parseDouble;

import android.content.ContentProviderOperation;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.OperationApplicationException;
import android.database.Cursor;
import android.location.Location;
import android.os.RemoteException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.Gson;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.res.StringRes;
import org.bspb.smartbirds.pro.R;
import org.bspb.smartbirds.pro.db.FormColumns;
import org.bspb.smartbirds.pro.db.MonitoringColumns;
import org.bspb.smartbirds.pro.db.SmartBirdsProvider;
import org.bspb.smartbirds.pro.enums.EntryType;
import org.bspb.smartbirds.pro.repository.TrackingRepository;
import org.bspb.smartbirds.pro.room.Tracking;
import org.bspb.smartbirds.pro.tools.SBGsonParser;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;
import java.util.UUID;

/**
 * Created by groupsky on 05.12.16.
 */

@EBean(scope = Singleton)
public class MonitoringManager {

    private static final DateFormat DATE_FORMATTER = new SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US);
    private static final Gson SERIALIZER = SBGsonParser.createParser();
    public static final String[] MONITORING_PROJECTION = {
            MonitoringColumns._ID,
            MonitoringColumns.STATUS,
            MonitoringColumns.DATA
    };
    private static final String[] ENTRY_PROJECTION = {
            FormColumns._ID,
            FormColumns.DATA
    };

    static {
        DATE_FORMATTER.setTimeZone(TimeZone.getTimeZone("UTC"));
    }

    @StringRes(R.string.tag_lat)
    static String tagLatitude;
    @StringRes(R.string.tag_lon)
    static String tagLongitude;

    ContentResolver contentResolver;
    Context rootContext;

    public MonitoringManager(Context context) {
        this.rootContext = context.getApplicationContext();
        this.contentResolver = rootContext.getContentResolver();
    }

    @NonNull
    private String generateMonitoringCode() {
        String uuid = UUID.randomUUID().toString();
        return String.format("%s-%s", DATE_FORMATTER.format(new Date()), uuid.substring(uuid.length() - 12));
    }

    @NonNull
    public Monitoring createNew() {
        String code = generateMonitoringCode();
        Monitoring monitoring = new Monitoring(code);
        monitoring.id = parseId(contentResolver.insert(SmartBirdsProvider.Monitorings.CONTENT_URI, toContentValues(monitoring)));
        return monitoring;
    }

    public void update(@NonNull Monitoring monitoring) {
        contentResolver.update(SmartBirdsProvider.Monitorings.withCode(monitoring.code), toContentValues(monitoring), null, null);
    }

    public void updateStatus(@NonNull String monitoringCode, @NonNull Monitoring.Status status) {
        ContentValues cv = new ContentValues();
        cv.put(MonitoringColumns.STATUS, status.name());
        contentResolver.update(SmartBirdsProvider.Monitorings.withCode(monitoringCode), cv, null, null);
    }

    public void updateStatus(@NonNull Monitoring monitoring, @NonNull Monitoring.Status status) {
        updateStatus(monitoring.code, status);
        monitoring.status = status;
    }

    private static ContentValues toContentValues(@NonNull Monitoring monitoring) {
        ContentValues cv = new ContentValues();
        cv.put(MonitoringColumns.CODE, monitoring.code);
        cv.put(MonitoringColumns.STATUS, monitoring.status.name());
        cv.put(MonitoringColumns.DATA, SERIALIZER.toJson(monitoring));
        return cv;
    }

    public static Monitoring monitoringFromCursor(@NonNull Cursor cursor) {
        Monitoring monitoring = SERIALIZER.fromJson(cursor.getString(cursor.getColumnIndexOrThrow(MonitoringColumns.DATA)), Monitoring.class);
        final int idIdx = cursor.getColumnIndex(MonitoringColumns._ID);
        if (idIdx != -1) {
            monitoring.id = cursor.getLong(idIdx);
        }
        final int statusIdx = cursor.getColumnIndex(MonitoringColumns.STATUS);
        if (statusIdx != -1) {
            monitoring.status = Monitoring.Status.valueOf(cursor.getString(statusIdx));
        }

        final int entriesIdx = cursor.getColumnIndex(MonitoringColumns.ENTRIES_COUNT);
        if (entriesIdx != -1) {
            monitoring.entriesCount = cursor.getInt(entriesIdx);
        }

        return monitoring;
    }

    public static MonitoringEntry entryFromCursor(@NonNull Cursor cursor) {
        MonitoringEntry entry = SERIALIZER.fromJson(cursor.getString(cursor.getColumnIndexOrThrow(FormColumns.DATA)), MonitoringEntry.class);
        final int idx = cursor.getColumnIndex(FormColumns._ID);
        if (idx != -1) {
            entry.id = cursor.getLong(idx);
        }
        return entry;
    }

    public Monitoring getMonitoring(@NonNull String monitoringCode) {
        Cursor cursor = contentResolver.query(SmartBirdsProvider.Monitorings.withCode(monitoringCode), MONITORING_PROJECTION, null, null, null);
        if (cursor != null) try {
            return cursor.moveToFirst() ? monitoringFromCursor(cursor) : null;
        } finally {
            cursor.close();
        }
        return null;
    }

    public Monitoring getActiveMonitoring() {
        Cursor cursor = contentResolver.query(SmartBirdsProvider.Monitorings.CONTENT_URI, MONITORING_PROJECTION, MonitoringColumns.STATUS + "=?", new String[]{wip.name()}, MonitoringColumns._ID + " desc");
        if (cursor == null) return null;
        try {
            if (!cursor.moveToFirst()) return null;
            return monitoringFromCursor(cursor);
        } finally {
            cursor.close();
        }
    }

    public Iterable<String> monitoringCodesForStatus(Monitoring.Status status) {
        Cursor cursor = contentResolver.query(SmartBirdsProvider.Monitorings.CONTENT_URI,
                new String[]{MonitoringColumns.CODE},
                MonitoringColumns.STATUS + "=?", new String[]{status.name()},
                MonitoringColumns._ID);
        if (cursor == null) return new ArrayList<>(0);
        try {
            List<String> monitoringCodes = new ArrayList<>(cursor.getCount());
            int codeIdx = cursor.getColumnIndexOrThrow(MonitoringColumns.CODE);
            for (cursor.moveToFirst(); !cursor.isAfterLast(); cursor.moveToNext())
                monitoringCodes.add(cursor.getString(codeIdx));
            return monitoringCodes;
        } finally {
            cursor.close();
        }
    }

    public int countMonitoringsForStatus(Monitoring.Status status) {
        int count = 0;
        try {
            Cursor cursor = contentResolver.query(SmartBirdsProvider.Monitorings.CONTENT_URI,
                    new String[]{MonitoringColumns.CODE},
                    MonitoringColumns.STATUS + "=?", new String[]{status.name()},
                    MonitoringColumns._ID);
            if (cursor != null)
                try {
                    count += cursor.getCount();
                } finally {
                    cursor.close();
                }
        } catch (Throwable t) {
            logException(t);
        }

        return count;
    }

    @Nullable
    public Monitoring getPausedMonitoring() {
        Cursor cursor = contentResolver.query(SmartBirdsProvider.Monitorings.CONTENT_URI, MONITORING_PROJECTION, MonitoringColumns.STATUS + "=?", new String[]{paused.name()}, MonitoringColumns._ID + " desc");
        if (cursor == null) return null;
        try {
            if (!cursor.moveToFirst()) return null;
            return monitoringFromCursor(cursor);
        } finally {
            cursor.close();
        }
    }

    public void deleteMonitoring(String monitoringCode) {
        ArrayList<ContentProviderOperation> ops = new ArrayList<>();
        ops.add(ContentProviderOperation.newDelete(SmartBirdsProvider.Forms.withMonitoringCode(monitoringCode)).build());
        ops.add(ContentProviderOperation.newDelete(SmartBirdsProvider.Monitorings.withCode(monitoringCode)).build());
        try {
            contentResolver.applyBatch(SmartBirdsProvider.AUTHORITY, ops);
        } catch (RemoteException | OperationApplicationException e) {
            logException(e);
        }
    }
}

File: app/src/main/kotlin/org/bspb/smartbirds/pro/utils/MonitoringManager.kt
package org.bspb.smartbirds.pro.utils

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.switchMap
import org.bspb.smartbirds.pro.R
import org.bspb.smartbirds.pro.SmartBirdsApplication
import org.bspb.smartbirds.pro.content.Monitoring
import org.bspb.smartbirds.pro.content.MonitoringEntry
import org.bspb.smartbirds.pro.content.TrackingLocation
import org.bspb.smartbirds.pro.enums.EntryType
import org.bspb.smartbirds.pro.repository.FormRepository
import org.bspb.smartbirds.pro.repository.MonitoringRepository
import org.bspb.smartbirds.pro.room.Form
import org.bspb.smartbirds.pro.room.MonitoringModel
import org.bspb.smartbirds.pro.room.Tracking
import org.bspb.smartbirds.pro.tools.SBGsonParser
import java.nio.charset.StandardCharsets
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class MonitoringManager private constructor(val context: Context) {

    companion object {
        private const val TAG = SmartBirdsApplication.TAG + ".MonitoringManager"

        @SuppressLint("StaticFieldLeak")
        @Volatile
        private var INSTANCE: MonitoringManager? = null

        private val DATE_FORMATTER: DateFormat = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US)
        private val SERIALIZER = SBGsonParser.createParser()

        fun getInstance(): MonitoringManager {
            checkNotNull(INSTANCE) { "MonitoringManager instance is null. init(Context context) must be called before getting the instance." }
            return INSTANCE!!
        }

        fun init(context: Context) {
            if (INSTANCE != null) {
                return
            }

            synchronized(this) {
                INSTANCE = MonitoringManager(context)
                INSTANCE
            }
        }

        fun entryFromDb(form: Form): MonitoringEntry {
            val entry = SERIALIZER.fromJson(
                String(form.data),
                MonitoringEntry::class.java
            )
            entry.id = form.id.toLong()
            return entry
        }

        fun monitoringFromDb(dbModel: MonitoringModel?): Monitoring? {
            dbModel ?: return null

            val monitoring = SERIALIZER.fromJson(
                String(dbModel.data!!),
                Monitoring::class.java
            )
            monitoring.id = dbModel.id
            monitoring.status = monitoring.status

            return monitoring
        }
    }

    private var tagLatitude: String? = null
    private var tagLongitude: String? = null
    private val formsRepository: FormRepository
    private val monitoringRepository: MonitoringRepository

    init {
        DATE_FORMATTER.timeZone = TimeZone.getTimeZone("UTC")
        tagLatitude = context.getString(R.string.tag_lat)
        tagLongitude = context.getString(R.string.tag_lon)
        formsRepository = FormRepository()
        monitoringRepository = MonitoringRepository()
    }

    suspend fun newEntry(monitoring: Monitoring, entryType: EntryType, data: HashMap<String?, String?>) {
        val entry = MonitoringEntry(monitoring.code, entryType)
        entry.data.putAll(data)
        formsRepository.insertForm(toDbModel(entry))
    }

    suspend fun updateEntry(
        monitoringCode: String,
        entryId: Long,
        entryType: EntryType,
        data: HashMap<String?, String?>
    ) {
        val entry = MonitoringEntry(monitoringCode, entryType)
        entry.data.putAll(data)
        entry.id = entryId

        formsRepository.updateForm(toDbModel(entry))
    }

    fun newTracking(monitoring: Monitoring, location: Location?): TrackingLocation? {
        /* TODO Enable tracking db persistence if needed
          Pause tracking db persistence since it is never read
         */
//        val repo = TrackingRepository()
//        repo.insertNewTracking(toDbModel(l))
        return TrackingLocation(monitoring.code, location!!)
    }

    suspend fun deleteLastEntry(monitoring: Monitoring) {
        val deletedRows = formsRepository.deleteLastEntry(monitoring.code)
        if (deletedRows.equals(0)) {
            Log.e(TAG, "could not delete last monitoring entry")
        }
    }

    suspend fun deleteEntries(ids: LongArray) {
        formsRepository.deleteLastEntries(ids)
    }

    fun getEntries(monitoring: Monitoring, entryType: EntryType): List<Form> {
        return formsRepository.getEntries(monitoring.code, entryType.name)
    }

    suspend fun createNew(): Monitoring {
        val code: String = generateMonitoringCode()
        val monitoring = Monitoring(code)
        monitoring.id = monitoringRepository.createMonitoring(toDbModel(monitoring))
        return monitoring
    }

    suspend fun update(monitoring: Monitoring) {
        monitoringRepository.updateMonitoring(toDbModel(monitoring))
    }

    suspend fun updateStatus(monitoringCode: String, status: Monitoring.Status) {
        monitoringRepository.updateStatus(monitoringCode, status)
    }

    suspend fun updateStatus(monitoring: Monitoring, status: Monitoring.Status) {
        updateStatus(monitoring.code, status)
        monitoring.status = status
    }

    suspend fun getMonitoring(monitoringCode: String): Monitoring? {
        return monitoringFromDb(monitoringRepository.getMonitoring(monitoringCode))
    }

    fun getMonitoringLive(monitoringCode: String): LiveData<Monitoring?> {
        return monitoringRepository.getMonitoringLive(monitoringCode).switchMap {
            liveData {
                emit(monitoringFromDb(it))
            }
        }
    }

    suspend fun getActiveMonitoring(): Monitoring? {
        return monitoringFromDb(monitoringRepository.getActiveMonitoring())
    }

    suspend fun monitoringCodesForStatus(status: Monitoring.Status): Iterable<String> {
        return monitoringRepository.getMonitoringCodesForStatus(status)
    }

    suspend fun getPausedMonitoring(): Monitoring? {
        return monitoringFromDb(monitoringRepository.getPausedMonitoring())
    }

    suspend fun deleteMonitoring(monitoringCode: String?) {
        monitoringRepository.deleteMonitoring(monitoringCode!!)
    }

    suspend fun countMonitoringsForStatus(status: Monitoring.Status): Int {
        return monitoringRepository.countMonitoringsForStatus(status)
    }

    private fun generateMonitoringCode(): String {
        val uuid = UUID.randomUUID().toString()
        return String.format("%s-%s", DATE_FORMATTER.format(Date()), uuid.substring(uuid.length - 12))
    }

    private fun toDbModel(monitoring: Monitoring): MonitoringModel {
        return MonitoringModel(
            monitoring.id,
            monitoring.code,
            monitoring.status.name,
            SERIALIZER.toJson(monitoring).toByteArray(StandardCharsets.UTF_8)
        )
    }

    private fun toDbModel(entry: MonitoringEntry): Form {
        return Form(
            entry.id,
            entry.monitoringCode,
            entry.type.name,
            entry.data[tagLatitude]!!.toDouble(),
            entry.data[tagLongitude]!!.toDouble(),
            SERIALIZER.toJson(entry).toByteArray(StandardCharsets.UTF_8)
        )
    }

    private fun toDbModel(location: TrackingLocation): Tracking {
        return Tracking(
            location.id,
            location.monitoringCode,
            location.time,
            location.latitude,
            location.longitude,
            location.altitude
        )
    }
}